File: wellnest-frontend/src/services/trainerService.js

const TRAINERS = [
    {
        id: 't1',
        name: 'Alex Johnson',
        specialties: ['Muscle Gain', 'Strength Training'],
        experience: 8,
        rating: 4.9,
        location: 'New York',
        availability: ['Mon', 'Wed', 'Fri'],
        bio: 'Certified strength coach helping you build muscle and confidence.',
        image: 'https://images.unsplash.com/photo-1567013127542-490d757e51fc?q=80&w=1887&auto=format&fit=crop',
        email: 'alex.j@wellnest_trainers.com',
        phone: '+1 (555) 012-3456'
    },
    {
        id: 't2',
        name: 'Sarah Lee',
        specialties: ['Yoga', 'Flexibility', 'Mental Wellness'],
        experience: 5,
        rating: 4.8,
        location: 'Online',
        availability: ['Tue', 'Thu', 'Sat'],
        bio: 'Yoga instructor focused on holistic health and mindfulness.',
        image: 'https://images.unsplash.com/photo-1518611012118-696072aa579a?q=80&w=2070&auto=format&fit=crop',
        email: 'sarah.yoga@wellnest_trainers.com',
        phone: '+1 (555) 012-7890'
    },
    {
        id: 't3',
        name: 'Mike Chen',
        specialties: ['Weight Loss', 'HIIT', 'Cardio'],
        experience: 6,
        rating: 4.7,
        location: 'San Francisco',
        availability: ['Mon', 'Tue', 'Thu', 'Fri'],
        bio: 'High energy trainer specialized in fat loss and metabolic conditioning.',
        image: 'https://images.unsplash.com/photo-1627931105822-6b99d5543c3a?u=4r&q=80&w=2000&auto=format&fit=crop',
        email: 'mike.chen@wellnest_trainers.com',
        phone: '+1 (555) 012-4567'
    },
    {
        id: 't4',
        name: 'Jessica Davis',
        specialties: ['Rehabilitation', 'Mobility', 'Senior Fitness'],
        experience: 12,
        rating: 5.0,
        location: 'Chicago',
        availability: ['Wed', 'Fri'],
        bio: 'Helping you recover from injuries and move pain-free.',
        image: 'https://images.unsplash.com/photo-1571019614242-c5c5dee9f50b?q=80&w=2070&auto=format&fit=crop',
        email: 'jessica.rehab@wellnest_trainers.com',
        phone: '+1 (555) 012-8901'
    },
    {
        id: 't5',
        name: 'David Kim',
        specialties: ['CrossFit', 'Endurance', 'Muscle Gain'],
        experience: 4,
        rating: 4.6,
        location: 'Online',
        availability: ['Sat', 'Sun'],
        bio: 'Pushing your limits with functional fitness and endurance training.',
        image: 'https://images.unsplash.com/photo-1594381898411-846e7d193883?q=80&w=1887&auto=format&fit=crop',
        email: 'david.fit@wellnest_trainers.com',
        phone: '+1 (555) 012-2345'
    }
];

export const getAllTrainers = () => {
    return TRAINERS;
};

export const matchTrainers = (criteria) => {
    return TRAINERS.filter(trainer => {
        // Match Goal,if selected, check if any specialty matches vaguely or exactly
        let goalMatch = true;
        if (criteria.goal && criteria.goal !== 'All') {
            goalMatch = trainer.specialties.some(s => s.toLowerCase().includes(criteria.goal.toLowerCase()));
            if (!goalMatch && criteria.goal === 'Weight Loss') {
                goalMatch = trainer.specialties.some(s => ['HIIT', 'Cardio'].includes(s));
            }
        }

        // Match Location,Online matches everyone if user accepts online, specific city otherwise
        let locationMatch = true;
        if (criteria.location && criteria.location !== 'Any') {
            if (criteria.location === 'Online') {
                locationMatch = trainer.location === 'Online';
            } else {
                locationMatch = trainer.location.toLowerCase() === criteria.location.toLowerCase() || trainer.location === 'Online';
            }
        }

        return goalMatch && locationMatch;
    });
};

File: src/main/java/com/wellnest/app/service/TrainerService.java
package com.wellnest.app.service;

import com.wellnest.app.dto.TrainerResponse;
import com.wellnest.app.model.Trainer;
import com.wellnest.app.repository.TrainerRepository;
import jakarta.annotation.PostConstruct;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class TrainerService {

    private final TrainerRepository trainerRepository;

    public TrainerService(TrainerRepository trainerRepository) {
        this.trainerRepository = trainerRepository;
    }

    // Initialize default trainers if database is empty
    @PostConstruct
    @Transactional
    public void initializeDefaultTrainers() {
        if (trainerRepository.count() == 0) {
            List<Trainer> defaultTrainers = Arrays.asList(
                    new Trainer(
                            "Alex Johnson",
                            Arrays.asList("Muscle Gain", "Strength Training"),
                            8,
                            4.9,
                            "New York",
                            Arrays.asList("Mon", "Wed", "Fri"),
                            "Certified strength coach helping you build muscle and confidence.",
                            "https://images.unsplash.com/photo-1567013127542-490d757e51fc?q=80&w=1887&auto=format&fit=crop",
                            "alex.j@wellnest_trainers.com",
                            "+1 (555) 012-3456"),
                    new Trainer(
                            "Sarah Lee",
                            Arrays.asList("Yoga", "Flexibility", "Mental Wellness"),
                            5,
                            4.8,
                            "Online",
                            Arrays.asList("Tue", "Thu", "Sat"),
                            "Yoga instructor focused on holistic health and mindfulness.",
                            "https://images.unsplash.com/photo-1518611012118-696072aa579a?q=80&w=2070&auto=format&fit=crop",
                            "sarah.yoga@wellnest_trainers.com",
                            "+1 (555) 012-7890"),
                    new Trainer(
                            "Mike Chen",
                            Arrays.asList("Weight Loss", "HIIT", "Cardio"),
                            6,
                            4.7,
                            "San Francisco",
                            Arrays.asList("Mon", "Tue", "Thu", "Fri"),
                            "High energy trainer specialized in fat loss and metabolic conditioning.",
                            "https://images.unsplash.com/photo-1627931105822-6b99d5543c3a?u=4r&q=80&w=2000&auto=format&fit=crop",
                            "mike.chen@wellnest_trainers.com",
                            "+1 (555) 012-4567"),
                    new Trainer(
                            "Jessica Davis",
                            Arrays.asList("Rehabilitation", "Mobility", "Senior Fitness"),
                            12,
                            5.0,
                            "Chicago",
                            Arrays.asList("Wed", "Fri"),
                            "Helping you recover from injuries and move pain-free.",
                            "https://images.unsplash.com/photo-1571019614242-c5c5dee9f50b?q=80&w=2070&auto=format&fit=crop",
                            "jessica.rehab@wellnest_trainers.com",
                            "+1 (555) 012-8901"),
                    new Trainer(
                            "David Kim",
                            Arrays.asList("CrossFit", "Endurance", "Muscle Gain"),
                            4,
                            4.6,
                            "Online",
                            Arrays.asList("Sat", "Sun"),
                            "Pushing your limits with functional fitness and endurance training.",
                            "https://images.unsplash.com/photo-1594381898411-846e7d193883?q=80&w=1887&auto=format&fit=crop",
                            "david.fit@wellnest_trainers.com",
                            "+1 (555) 012-2345"));

            trainerRepository.saveAll(defaultTrainers);
        }
    }

    public List<TrainerResponse> getAllTrainers() {
        return trainerRepository.findAllByOrderByRatingDesc()
                .stream()
                .map(this::toResponse)
                .collect(Collectors.toList());
    }

    public List<TrainerResponse> matchTrainers(String goal, String location) {
        List<Trainer> allTrainers = trainerRepository.findAll();

        return allTrainers.stream()
                .filter(trainer -> matchesGoal(trainer, goal))
                .filter(trainer -> matchesLocation(trainer, location))
                .map(this::toResponse)
                .collect(Collectors.toList());
    }

    private boolean matchesGoal(Trainer trainer, String goal) {
        if (goal == null || goal.equalsIgnoreCase("All")) {
            return true;
        }

        // Check if any specialty matches the goal
        boolean directMatch = trainer.getSpecialties().stream()
                .anyMatch(s -> s.toLowerCase().contains(goal.toLowerCase()));

        // Fallback logic: if goal is 'Weight Loss' and trainer has 'HIIT' or 'Cardio',
        // that's a match too
        if (!directMatch && goal.equalsIgnoreCase("Weight Loss")) {
            directMatch = trainer.getSpecialties().stream()
                    .anyMatch(s -> s.equalsIgnoreCase("HIIT") || s.equalsIgnoreCase("Cardio"));
        }

        return directMatch;
    }

    private boolean matchesLocation(Trainer trainer, String location) {
        if (location == null || location.equalsIgnoreCase("Any")) {
            return true;
        }

        if (location.equalsIgnoreCase("Online")) {
            return trainer.getLocation().equalsIgnoreCase("Online");
        }

        // If specific city, match city OR Online trainers are also available
        return trainer.getLocation().equalsIgnoreCase(location) ||
                trainer.getLocation().equalsIgnoreCase("Online");
    }

    public TrainerResponse getTrainerById(Long id) {
        return trainerRepository.findById(id)
                .map(this::toResponse)
                .orElse(null);
    }

    private TrainerResponse toResponse(Trainer trainer) {
        return new TrainerResponse(
                trainer.getId(),
                trainer.getName(),
                trainer.getSpecialties(),
                trainer.getExperience(),
                trainer.getRating(),
                trainer.getLocation(),
                trainer.getAvailability(),
                trainer.getBio(),
                trainer.getImage(),
                trainer.getEmail(),
                trainer.getPhone());
    }
}
